shotcut: 21.09.20 -> 23.11.29

uku3lig opened the pull request:
## Description of changes

Updates shotcut to the most recent version, 23.09.29, also making the switch from qmake to cmake, and from Qt5 to Qt6.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

uku3lig committed:
shotcut: 21.09.20 -> 23.11.29

SuperSandro2000 commented on pkgs/applications/video/shotcut/default.nix:
```diff
@@ -1,57 +1,52 @@
-assert lib.versionAtLeast mlt.version "6.24.0";
-
-mkDerivation rec {
+assert lib.versionAtLeast mlt.version "7.18.0";
```
I think we can remove this, we are probably not downgrading malt.

eclairevoyant commented on pkgs/applications/video/shotcut/default.nix:
```diff
@@ -62,18 +57,18 @@ mkDerivation rec {
+
   qtWrapperArgs = [
     "--prefix FREI0R_PATH : ${frei0r}/lib/frei0r-1"
     "--prefix LADSPA_PATH : ${ladspaPlugins}/lib/ladspa"
```
Was this actually tested? all of the previous PRs have the missing plugin regression mentioned in https://github.com/NixOS/nixpkgs/pull/201905#discussion_r1320980705 which doesn't seem to have been addressed here.

uku3lig replied:
thanks for pointing this out, I didn't notice it while testing. frei0r video effects all seem to be present, but ladspa audio filters are indeed missing.

i tried switching LADSPA_PATH to `--set` (from `--prefix`) since other video editing software in nixpkgs seems to use that instead, but sadly without results

woffs replied:
I think the reason is `melt` not recognizing the filters: they do not show up in `melt -query filters`, and there should be for example `melt -query filter=ladspa.1890`, which is not. I'm trying to compare with older versions where it  worked but I did not succeed yet (maybe because of glibc/libmltjackrack/libjack runtime linking insanities).

woffs replied:
It was apparently broken by 5f702e92207a43e189a180950f6dc6f8c4d22cfd.
`jackrack/filter_ladspa.yml` is just missing from mlt.

The LADSPA filters reappear and work if mlt is overridden with
`shotcut = qt6Packages.callPackage ../applications/video/shotcut { mlt = mlt.override { enableQt = true; }; };`
but this pulls Qt5 back in, which is not what we want, I guess.
So the fix should be done in mlt package, replacing qt5 with qt6 there and enabling it for shotcut.

Scrumplex committed:
mlt: fix jackrack support

This makes ladspa filters work. Apparently glib is required for jackrack
filters to function.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

uku3lig commented:
@eclairevoyant @woffs can you try again with this new commit? it seems that adding glib to mlt fixes this problem, at the very least it seems to work on my machine just fine

woffs commented:
> @eclairevoyant @woffs can you try again with this new commit? it seems that adding glib to mlt fixes this problem, at the very least it seems to work on my machine just fine

FAME!! :+1:

woffs commented on pkgs/applications/video/shotcut/default.nix:
```diff
@@ -1,79 +1,73 @@
+    "--set FREI0R_PATH ${frei0r}/lib/frei0r-1"
+    "--set LADSPA_PATH ${ladspaPlugins}/lib/ladspa"
+    "--prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [jack1 SDL2]}"
     "--prefix PATH : ${mlt}/bin"
```
mlt in PATH is not needed anymore, because we substitute the full path above

uku3lig replied:
ok it works, maybe i just ran the wrong version or forgot to rebuild

woffs commented on pkgs/applications/video/shotcut/default.nix:
```diff
@@ -1,79 +1,73 @@
+, qtcharts
+, cmake
 , gitUpdater
+,
```
one too many ;-)

eclairevoyant commented on pkgs/applications/video/shotcut/default.nix:
```diff
@@ -1,79 +1,73 @@
 
+  patches = [
+    # remove when updating to 23.11
+    ./dont-shadow-final-properties.patch
```
I don't understand the patch, especially since we are basically beyond 23.11 by now. Please add a comment explianing its purpose, and if possible, fetch from upstream.

uku3lig replied:
i should've been more precise with my comment, i was talking about shotcut 23.11, not nixos! the patch fixes the timeline not being rendered on qt 6.6, i'll add a comment for that later

eclairevoyant replied:
Can it be fetched instead of vendoring it?

eclairevoyant reviewed (changes requested):
Other than the patch, can we set `meta.mainProgram`? Since you're already updating this file substantially

SuperSandro2000 commented on pkgs/applications/video/shotcut/default.nix:
```diff
@@ -92,6 +85,7 @@ mkDerivation rec {
     homepage = "https://shotcut.org";
     license = licenses.gpl3Plus;
     maintainers = with maintainers; [ goibhniu woffs peti ];
+    mainProgram = "shotcut";
```
Please rebase onto master where this already exists.

SuperSandro2000 commented on pkgs/applications/video/shotcut/default.nix:
```diff
@@ -1,78 +1,71 @@
+    "-DSHOTCUT_VERSION=${version}"
   ];
 
   prePatch = ''
```
This breaks patches and should be done in postPatch. Also why are those not patches? That would make silent breakages way less likely.

uku3lig replied:
i merely fixed/upgraded what was already present, but yeah it's probably a better idea to use a proper patch file

woffs commented on pkgs/applications/video/shotcut/default.nix:
```diff
@@ -1,78 +1,71 @@
-    "--prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ jack1 SDL2 ]}"
-    "--prefix PATH : ${mlt}/bin"
+  patches = [
+    # remove when updating to shotcut 23.11
```
btw. https://github.com/mltframework/shotcut/releases/tag/v23.11.29 :-)

uku3lig replied:
i saw, the only problem is that it requires mlt 7.22.0, which is not yet in nixpkgs, see #270928

eclairevoyant replied:
That PR could be combined here specifically bc it's a hard dependency

uku3lig replied:
i think i'm just gonna keep it to 23.09.29 for now, unless the mlt pr gets merged before this one, in which case i'll update to 23.11.29

eclairevoyant replied:
Less PRs is better given the limited number of committers - 3 PRs becomes just 1 if you update accordingly. I'd prefer to combine the `mlt` update here.

uku3lig replied:
the pr got merged! updated to 23.11.29

uku3lig commented:
the mlt 7.22.0 PR got merged, so updated this pr to 23.11.29

shotcut seems to be working fine on my end (with no missing audio filters!), lmk if you find any issues

MaxwellDupre reviewed (approved):
Changes look good to me.

pbsds commented:
Result of `nixpkgs-review pr 269444` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>flowblade</li>
    <li>hikounomizu</li>
    <li>libsForQt5.kdenlive (plasma5Packages.kdenlive)</li>
    <li>libsForQt5.mlt (plasma5Packages.mlt)</li>
    <li>libsForQt5.mlt.dev (plasma5Packages.mlt.dev)</li>
    <li>mlt</li>
    <li>mlt.dev</li>
    <li>python310Packages.mlt</li>
    <li>python310Packages.mlt.dev</li>
    <li>python311Packages.mlt</li>
    <li>python311Packages.mlt.dev</li>
    <li>shotcut</li>
    <li>synfigstudio</li>
  </ul>
</details>

pbsds commented:
Thanks!
